### ISSUE TITLE
Cannot use Zod .transform

### ISSUE THREAD

**Issue by alii** [react-hook-form/resolvers#416]
**Describe the bug**
Unable to use `.transform` with the zod resolver

**To Reproduce**
Use an input and see an error
<img width="1160" alt="image" src="https://user-images.githubusercontent.com/25351731/173365481-43d02615-9ed5-40f1-8ef9-0897e12f9e65.png">

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-zod-resolver-example-forked-khj93c?file=/src/App.tsx

**Expected behavior**
The transform happens as expected, and errors thrown during transformation are treated as form errors

**Desktop (please complete the following information):**
- OS: MacOS 13 Ventura
- Browser: Chrome 102.0.5005.61 (Official Build) (arm64) 

**Additional context**
Replaces #415 (Error carried forward)

**Comment by bluebill1049** [comment]
<img width="816" alt="Screen Shot 2022-06-15 at 12 45 35 pm" src="https://user-images.githubusercontent.com/10513364/173725793-6335da05-a990-405c-92f7-5975d85d1610.png">

This is the error log that I am getting.

**Comment by bluebill1049** [comment]
any update on this issue? as you can see in the console the error occurred at the zod, which indicates to me the schema could be invalid.

**Comment by Lykan9999** [comment]
> <img alt="Screen Shot 2022-06-15 at 12 45 35 pm" width="816" src="https://user-images.githubusercontent.com/10513364/173725793-6335da05-a990-405c-92f7-5975d85d1610.png">
> 
> This is the error log that I am getting.

I am getting this error when using `preprocess` in the zod schema. easiest solution (not optimal) was  to drop its usage

**Comment by jorisre** [comment]
Please open a new issue thank you

**Comment by alexgleason** [comment]
Same issue here. This does not work:

```ts
const fileListSchema = z.custom<FileList>((val) => val instanceof FileList);

const schema = z.object({
  avatar: fileListSchema.transform((fileList) => fileList.item(0)), // This does not work
});

const { register, handleSubmit, getValues, formState } = useForm<z.infer<typeof schema>>({
  resolver: zodResolver(schema),
});
```

`avatar` comes back as a `FileList` even though the schema transforms it into a `File`

**Comment by jorisre** [comment]
@alexgleason Please open a new issue

**Comment by joaopedrodcf** [comment]
@alexgleason @jorisre 

I think I found a solution by looking into the useForm generics definition:

As we can see we have three generics being passed:
![image](https://github.com/react-hook-form/resolvers/assets/11476152/2ed5cb8b-e03f-46c5-9aac-8e702600172c)

```ts
const fileListSchema = z.custom<FileList>((val) => val instanceof FileList);

const schema = z.object({
  avatar: fileListSchema.transform((fileList) => fileList.item(0)), // This does not work
});

const { register, handleSubmit, getValues, formState } = useForm<z.input<typeof schema>, object, z.output<typeof schema>>({
  resolver: zodResolver(schema),
});
```

Basically you can defined input, context and output as the generics 

Docs of zod: https://zod.dev/?id=type-inference 
TLDR: z.output is equal to z.infer

This way when you defined defautValues with these generics you get the type before transform and when you call handleSubmit it will be after transform.

Hopefully this helps others.

**Comment by LeonardBoedi** [comment]
@joaopedrodcf 

You just saved me a lot of pain. Thank you!

**Comment by kainbryanjones** [comment]
> [@alexgleason](https://github.com/alexgleason) [@jorisre](https://github.com/jorisre)
> 
> I think I found a solution by looking into the useForm generics definition:
> 
> As we can see we have three generics being passed: ![image](https://github.com/react-hook-form/resolvers/assets/11476152/2ed5cb8b-e03f-46c5-9aac-8e702600172c)
> 
> const fileListSchema = z.custom<FileList>((val) => val instanceof FileList);
> 
> const schema = z.object({
>   avatar: fileListSchema.transform((fileList) => fileList.item(0)), // This does not work
> });
> 
> const { register, handleSubmit, getValues, formState } = useForm<z.input<typeof schema>, object, z.output<typeof schema>>({
>   resolver: zodResolver(schema),
> });
> Basically you can defined input, context and output as the generics
> 
> Docs of zod: https://zod.dev/?id=type-inference TLDR: z.output is equal to z.infer
> 
> This way when you defined defautValues with these generics you get the type before transform and when you call handleSubmit it will be after transform.
> 
> Hopefully this helps others.

I am doing this but getting an TS error. I can see in the types for useForm that the TTransformedValues generic type is never passed to the UseFormProps, thus the type is giving my this error. What to do here? It wants the output to be the same as the input.

<img width="1074" alt="Image" src="https://github.com/user-attachments/assets/0dbe99ef-9cf7-4397-b4b9-a2c4dd9f0730" />

**Comment by jorisre** [comment]
@kainbryanjones There is an ongoing PR #753 to fix this issue.
